SF Salaries Exercise

Welcome to a quick exercise for you to practice your pandas skills! We will be using the SF Salaries Dataset from Kaggle! Just follow along and complete the tasks outlined in bold below. The tasks will get harder and harder as you go along.


1. Read Salaries.csv as a dataframe called sal.

In [13]:
import pandas as pd
df=pd.read_csv("Salaries.csv")
print(df)

            Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148

/var/folders/ds/6kfbb9qd0mx8352l0y59g3tr0000gn/T/ipykernel_64613/3532366739.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


2.Check the head of the DataFrame.
3.Use the .info() method to find out how many entries there are

In [16]:
print(df.head(1))
print(df.info())



   Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  

4.What is the average BasePay ?
a. 66325.44884050643

In [17]:
df['BasePay'].mean()

66325.4488404877

5.What is the highest amount of OvertimePay in the dataset ?
a.245131.88

In [19]:
df["OvertimePay"].max()

245131.88

6.What is the job title of JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). 
a.24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

In [18]:
df[df['EmployeeName']=='JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

7. How much does JOSEPH DRISCOLL make (including benefits)?
a.24    270324.91
Name: TotalPayBenefits, dtype: float64

In [21]:
df[df['EmployeeName']=='JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

8.What is the name of highest paid person (including benefits)?
a.
Id	EmployeeName	JobTitle	BasePay	OvertimePay	OtherPay	Benefits	TotalPay	TotalPayBenefits	Year	Notes	Agency	Status
0	1	NATHANIEL FORD	GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY	167411.18	0.0	400184.25	NaN	567595.43	567595.43	2011	NaN	San Francisco	NaN


In [20]:
df[df['TotalPayBenefits']==df['TotalPayBenefits'].max()]

Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       NaN  567595.43         567595.43   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN

9.What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?
a.
Id	EmployeeName	JobTitle	BasePay	OvertimePay	OtherPay	Benefits	TotalPay	TotalPayBenefits	Year	Notes	Agency	Status
148653	148654	Joe Lopez	Counselor, Log Cabin Ranch	0.0	0.0	-618.13	0.0	-618.13	-618.13	2014	NaN	San Francisco	NaN

In [22]:
df[df['TotalPayBenefits']==df['TotalPayBenefits'].min()]

Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency  Status  
148653  San Francisco     NaN

10.What was the average (mean) BasePay of all employees per year? (2011-2014) ?
a.Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [23]:
x=df.groupby('Year')
print(type(df.Year))
x['BasePay'].mean()

<class 'pandas.core.series.Series'>


Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

11.How many unique job titles are there? 
a.2159

In [24]:
df['JobTitle'].nunique()

2159

12. What are the top 5 most common jobs? 
a.
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

In [35]:
df.JobTitle.value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

13.How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) 
a.202

In [33]:
sum(df[df['Year']==2013]['JobTitle'].value_counts()==1)

202

14.How many people have the word Chief in their job title? (Bonus)
a. 477

In [32]:
def chief(title):
    if 'chief' in title.lower():
        return True
    else:
        return False
sum(df['JobTitle'].apply(chief))

627

15. Bonus: Is there a correlation between length of the Job Title string and Salary?
a.
	title_len	TotalPayBenefits
title_len	1.000000	-0.036878
TotalPayBenefits	-0.036878	1.000000



In [25]:
df['titlelen']=df['JobTitle'].apply(len)
df[['titlelen','TotalPayBenefits']].corr()

titlelen  TotalPayBenefits
titlelen          1.000000         -0.036878
TotalPayBenefits -0.036878          1.000000

IndentationError: unexpected indent (895479935.py, line 2)